packages/langium/src/references/linker.ts

```diff
@@ -50,16 +72,15 @@ export class DefaultLinker implements Linker {
     }
 
     buildReference(node: AstNode, refNode: CstNode, text: string, crossRefId: string): Reference {
-        const link = this.link.bind(this);
-        const reference: Reference & { _ref?: AstNode } = {
+        const getLinkedNode = this.getLinkedNode.bind(this);
+        const reference: Reference & { _ref?: AstNode | LinkingError } = {
             $refNode: refNode,
             $refName: text,
             get ref() {
-                if (reference._ref === undefined || !getDocument(reference._ref).valid) {
-                    // TODO handle linking errors
-                    reference._ref = link(node, text, crossRefId);
+                if (reference._ref === undefined || isAstNode(reference._ref) && !getDocument(reference._ref).valid) {
```

Comment:
Should it be `or not isAstNode`?  `|| !isAstNode(reference._ref` ?

Reply:
@msujew
I think one of the main ideas to have a LinkingError (not just `undefined`) is to prevent trying to resolve the reference when it is clear it can not be resolved.
Otherwise validation and other code that operates on a linked document will try to link again and again